crates/roc_std/src/roc_list.rs

```diff
@@ -899,6 +916,64 @@ where
     }
 }
 
+#[repr(transparent)]
+pub struct ReadOnlyRocList<T>(RocList<T>)
+where
+    T: RocRefcounted;
+
+unsafe impl<T> Send for ReadOnlyRocList<T> where T: Send + RocRefcounted {}
+unsafe impl<T> Sync for ReadOnlyRocList<T> where T: Sync + RocRefcounted {}
+
+impl<T> RocRefcounted for ReadOnlyRocList<T>
+where
+    T: RocRefcounted,
+{
+    fn inc(&mut self) {
+    }
+
+    fn dec(&mut self) {
+    }
+
+    fn is_refcounted() -> bool {
+        true
+    }
+}
+
+impl<T> Clone for ReadOnlyRocList<T>
+where
+    T: Clone + RocRefcounted,
+{
+    fn clone(&self) -> Self {
+        ReadOnlyRocList(self.0.clone())
+    }
+}
+
+impl<T> From<RocList<T>> for ReadOnlyRocList<T>
+where
+    T: Clone + RocRefcounted,
+{
+    fn from(mut l: RocList<T>) -> Self {
+        if l.is_unique() {
+            unsafe {l.set_readonly()};
+        }
+        if l.is_readonly() {
+            ReadOnlyRocList(l)
+        } else {
+            // This is not unique, do a deep copy.
+            ReadOnlyRocList::from(RocList::from_slice(&l))
```

Comment:
Does calling the from here call `set_readonly` because it's now unique?

Reply:
exactly